geomesa-accumulo/geomesa-accumulo-datastore/src/test/scala/org/locationtech/geomesa/accumulo/data/AccumuloFeatureReaderTest.scala

```diff
@@ -84,7 +224,31 @@ class AccumuloFeatureReaderTest extends Specification with TestWithDataStore {
       count must beLessThan(10)
       stats must haveLength(1)
       stats.head must beAnInstanceOf[QueryStat]
-      stats.head.asInstanceOf[QueryStat].hits mustEqual 10
+      stats.head.asInstanceOf[QueryStat].hits must beLessThan(20L)
+    }
+
+    "be able to limit  features in bin results and collect stats through geoserver" in {
+      val stats = ArrayBuffer.empty[Stat]
+      val query = new Query(sftName, filter)
+      query.setMaxFeatures(10)
+
+      val collection = dataStoreWithStats(stats).getFeatureSource(sftName).getFeatures(query)
+
+      // put the hints in the request after getting the feature collection
+      // to mimic the bin output format workflow
+      val hints = Map(QueryHints.BIN_TRACK_KEY -> "name", QueryHints.BIN_BATCH_SIZE_KEY -> 10)
+      QueryPlanner.setPerThreadQueryHints(hints.asInstanceOf[Map[AnyRef, AnyRef]])
+
+      val reader = collection.features()
+
+      var count = 0
+      while (reader.hasNext) { reader.next(); count += 1 }
+      reader.close()
+
+      count must beLessThan(10)
```

Comment:
yeah, somewhat - the features that get returned depend on the ranges that get set up in accumulo, as each range will create a new bin iterator and return aggregated features for that range. Since each feature can have 0-10 bin records (based on the bin batch size hint), we don't know exactly how many features will get returned before the bin limit of 10 (based on maxFeatures) is hit. But conservatively, if there is at least one feature that contains 2+ bin records, there will be less than 10 features. I'll add a comment explaining it.